lib/livebook/session.ex

```diff
@@ -1666,25 +1670,57 @@ defmodule Livebook.Session do
   end
 
   defp set_runtime_secret(state, secret) do
-    secret = {"LB_#{secret.name}", secret.value}
-    Runtime.put_system_envs(state.data.runtime, [secret])
+    Runtime.put_system_envs(state.data.runtime, [set_env_prefix(secret)])
   end
 
   defp set_runtime_secrets(state, secrets) do
-    secrets = Enum.map(secrets, fn {name, value} -> {"LB_#{name}", value} end)
+    secrets = for secret <- secrets, do: set_env_prefix(secret)
     Runtime.put_system_envs(state.data.runtime, secrets)
   end
 
-  defp delete_runtime_secrets(state, secret_names) do
-    secret_names = Enum.map(secret_names, &"LB_#{&1}")
-    Runtime.delete_system_envs(state.data.runtime, secret_names)
-  end
-
   defp set_runtime_env_vars(state) do
-    env_vars = Enum.map(Livebook.Settings.fetch_env_vars(), &{&1.name, &1.value})
+    env_vars = Livebook.Settings.fetch_env_vars()
     Runtime.put_system_envs(state.data.runtime, env_vars)
   end
 
+  defp delete_runtime_secrets(state, secrets) do
+    secrets = for secret <- secrets, do: set_env_prefix(secret)
+    Runtime.delete_system_envs(state.data.runtime, secrets)
+  end
+
+  defp set_env_prefix(%EnvVar{} = env_var), do: env_var
+  defp set_env_prefix(%Secret{} = secret), do: %{secret | name: "LB_#{secret.name}"}
+
+  defp set_session_secret(state, %Secret{origin: :system_env} = secret) do
+    session_secrets = Enum.reject(state.session_secrets, &(&1.name == secret.name))
+    %{state | session_secrets: [secret | session_secrets]}
+  end
+
+  defp set_session_secret(state, _secret), do: state
+
+  defp unset_session_secret(state, %Secret{origin: :system_env} = secret) do
+    %{state | session_secrets: Enum.reject(state.session_secrets, &(&1.name == secret.name))}
+  end
+
+  defp unset_session_secret(state, _secret), do: state
+
+  defp maybe_set_another_secret(state, %Secret{} = secret) do
+    data_secrets = state.data.secrets
+    session_secrets = state.session_secrets
+    find_fun = &(&1.name == secret.name and &1.origin != secret.origin)
+
+    data_secret = Enum.find(data_secrets, find_fun)
+    session_secret = Enum.find(session_secrets, find_fun)
+
+    cond do
+      session_secret != nil -> set_runtime_secret(state, session_secret)
+      data_secret != nil -> set_runtime_secret(state, data_secret)
+      :else -> :ok
+    end
+
+    state
+  end
```

Comment:
I am not sure about the changes in this file. The session will only concern itself with `:system_env` origins, so I don't think it should ever receive the other types.